consensus/src/pipeline/virtual_processor/utxo_validation.rs

```diff
@@ -184,34 +185,64 @@ impl VirtualStateProcessor {
                 .par_iter() // We can do this in parallel without complications since block body validation already ensured
                             // that all txs within each block are independent
                 .skip(1) // Skip the coinbase tx.
-                .filter_map(|tx| self.validate_transaction_in_utxo_context(tx, &utxo_view, pov_daa_score))
+                .filter_map(|tx| self.validate_transaction_in_utxo_context(tx, &utxo_view, pov_daa_score).ok())
                 .collect()
         })
     }
 
-    /// Attempts to populate the transaction with UTXO entries and performs all tx validations
-    fn validate_transaction_in_utxo_context<'a>(
-        self: &Arc<Self>,
+    /// Attempts to populate the transaction with UTXO entries and performs all utxo-related tx validations
+    pub(super) fn validate_transaction_in_utxo_context<'a>(
+        &self,
         transaction: &'a Transaction,
         utxo_view: &impl UtxoView,
         pov_daa_score: u64,
-    ) -> Option<ValidatedTransaction<'a>> {
+    ) -> TxResult<ValidatedTransaction<'a>> {
         let mut entries = Vec::with_capacity(transaction.inputs.len());
         for input in transaction.inputs.iter() {
             if let Some(entry) = utxo_view.get(&input.previous_outpoint) {
                 entries.push(entry);
             } else {
-                return None; // Missing inputs
+                // Missing at least one input. For perf considerations, we report once a single miss is detected and avoid collecting all possible misses.
+                return Err(TxRuleError::MissingTxOutpoints);
             }
         }
         let populated_tx = PopulatedTransaction::new(transaction, entries);
         let res = self.transaction_validator.validate_populated_transaction_and_get_fee(&populated_tx, pov_daa_score);
         match res {
-            Ok(calculated_fee) => Some(populated_tx.to_validated(calculated_fee)),
+            Ok(calculated_fee) => Ok(ValidatedTransaction::new(populated_tx, calculated_fee)),
             Err(tx_rule_error) => {
                 info!("tx rule error {} for tx {}", tx_rule_error, transaction.id());
-                None
+                Err(tx_rule_error)
             }
         }
     }
+
+    /// Populates the mempool transaction with maximally found UTXO entry data and proceeds to validation if all found
+    pub(super) fn validate_mempool_transaction_in_utxo_context(
+        &self,
+        mutable_tx: &mut MutableTransaction,
+        utxo_view: &impl UtxoView,
+        pov_daa_score: u64,
+    ) -> TxResult<()> {
+        let mut has_missing_outpoints = false;
+        for i in 0..mutable_tx.inputs().len() {
+            if mutable_tx.entries[i].is_some() {
+                // We prefer a previously populated entry if such exists
```

Comment:
Something to think about: what if the DAA score changes and it affects coinbase maturity?

Reply:
Added a comment